apps/settings/js/panels/call_forwarding/panel.js

```diff
@@ -0,0 +1,248 @@
+
+  return function ctor_call_barring() {
+    return SettingsPanel({
+      onInit: function(panel, options) {
+        this._elements = {};
+        this._elements.panel = panel;
+
+        this._cardIndex = options.cardIndex || 0;
+        this._conns = window.navigator.mozMobileConnections;
+        this._conn = this._conns[this._cardIndex];
+
+        this._elements.liCfuDesc = panel.querySelector('#li-cfu-desc');
+        this._elements.liCfmbDesc = panel.querySelector('#li-cfmb-desc');
+        this._elements.liCfnRepDesc = panel.querySelector('#li-cfnrep-desc');
+        this._elements.liCfnReaDesc = panel.querySelector('#li-cfnrea-desc');
+
+        this._elements.cfuDesc = panel.querySelector('#cfu-desc');
+        this._elements.cfmbDesc = panel.querySelector('#cfmb-desc');
+        this._elements.cfnRepDesc = panel.querySelector('#cfnrep-desc');
+        this._elements.cfnReaDesc = panel.querySelector('#cfnrea-desc');
+
+        this._elements.mobileBusyMenuItem =
+          panel.querySelector('.callForwardingMobileBusy');
+        this._elements.noReplyMenuItem =
+          panel.querySelector('.callForwardingNoReply');
+        this._elements.notReachableMenuItem =
+          panel.querySelector('.callForwardingNotReachable');
+        this._elements.unConditionalMenuItem =
+          panel.querySelector('.callForwardingUnconditional');
+
+        this._bindMenuItemsClickEvent();
+      },
+
+      onBeforeShow: function(panel, options) {
+        if (options.cardIndex) {
+          this._cardIndex = options.cardIndex;
+          this._conn = this._conns[this._cardIndex];
+        }
+
+        this._updateSubpanels();
+      },
+
+      _updateAllDescInfoWithL10nId: function(l10nId) {
+        this._elements.cfuDesc.setAttribute('data-l10n-id', l10nId);
+        this._elements.cfmbDesc.setAttribute('data-l10n-id', l10nId);
+        this._elements.cfnRepDesc.setAttribute('data-l10n-id', l10nId);
+        this._elements.cfnReaDesc.setAttribute('data-l10n-id', l10nId);
+      },
+
+      _updateAllDescInfoWithRules: function(allRules) {
+        var mapping = {
+          'unconditional': 'cfuDesc',
+          'mobilebusy': 'cfmbDesc',
+          'noreply': 'cfnRepDesc',
+          'notreachable': 'cfnReaDesc'
+        };
+
+        for (var key in mapping) {
+          var rules = allRules[key];
+          var elementId = mapping[key];
+          var element = this._elements[elementId];
+          var rule = CallUtils.findActiveVoiceRule(rules);
+          if (rule) {
+            navigator.mozL10n.setAttributes(element,
+              'callForwardingForwardingVoiceToNumber', {
+                number: rule.number
+            });
+          } else {
+            element.setAttribute('data-l10n-id',
+              'callForwardingNotForwarding');
+          }
+        }
+      },
+
+      _bindMenuItemsClickEvent: function() {
+        // We will concate strings like (for grep use)
+        //
+        // ril.cf.mobilebusy.enabled
```

Comment:
Just realized that these keys were only used for presenting the current call forwarding settings on the dialog and panel. Now we can use an observable to achieve the same purpose without polluting the settings db. Let's remove all logic for updating this settings keys.